kernel/libs/xarray/src/lib.rs

```diff
@@ -0,0 +1,252 @@
+//! use crate::task::disable_preempt;
+//!
+//! let xarray_arc: XArray<Arc<i32>> = XArray::new();
+//! let value = Arc::new(10);
+//! xarray_arc.lock().store(10, value);
+//!
+//! let guard = disable_preempt();
+//! assert_eq!(*xarray_arc.load(&guard, 10).unwrap().as_ref(), 10);
+//!
+//! // Usage of the cursor
+//!
+//! let locked_xarray = xarray_arc.lock();
+//! let cursor_mut = locked_xarray.cursor_mut(100);
+//!
+//! let value = Arc::new(100);
+//! cursor_mut.store(value);
+//! assert_eq!(cursor_mut.load(10).unwrap().as_ref(), 100);
+//! let cursor = xarray_arc.cursor(&guard, 100);
+//! assert_eq!(cursor.load(10).unwrap().as_ref(), 100);
+//! ```
+//!
+//! # Background
+//!
+//! The XArray concept was originally introduced by Linux, which keeps the data structure of
+//! [Linux Radix Trees](https://lwn.net/Articles/175432/).
+
+#![no_std]
+#![deny(unsafe_code)]
+
+extern crate alloc;
+
+use core::marker::PhantomData;
+
+pub use cursor::{Cursor, CursorMut, SetMarkError};
+use entry::XEntry;
+use mark::NoneMark;
+pub use mark::XMark;
+use ostd::{
+    sync::{
+        non_null::NonNullPtr, LocalIrqDisabled, PreemptDisabled, RcuOption, SpinGuardian, SpinLock,
+        SpinLockGuard,
+    },
+    task::atomic_mode::AsAtomicModeGuard,
+};
+pub use range::Range;
+
+mod cursor;
+mod entry;
+mod mark;
+mod node;
+mod range;
+
+const BITS_PER_LAYER: usize = 6;
+const SLOT_SIZE: usize = 1 << BITS_PER_LAYER;
+const SLOT_MASK: usize = SLOT_SIZE - 1;
+
+/// A RCU-based `XArray` implementation.
+///
+/// `XArray` is used to store [`NonNullPtr`], with the additional requirement that user-stored
+/// pointers must have a minimum alignment of 2 bytes.
+///
+/// `XArray` is RCU-based, which means:
+/// - Multiple concurrent readers are permitted.
+/// - Only one writer is allowed at a time.
+/// - Simultaneous read operations are allowed while writing.
+/// - Readers may see stale data (see [`Cursor`] and [`CursorMut`] for more information).
+///
+/// Interaction with `XArray` is generally through `Cursor` and `CursorMut`. Similar to
+/// XArray's read-write properties, multiple `Cursor`s may coexist (shared read access) and
+/// only one `CursorMut` may exist at a time (exclusive write access).
+///
+/// To create a `Cursor`, users can invoke [`XArray::cursor`] with an atomic-guard.
+/// To create a `CursorMut`, users need to call [`XArray::lock`] or [`XArray::lock_irq_disabled`]
+/// first to obtain a [`LockedXArray`] first.
+///
+/// `XArray` enables marking of individual items for user convenience. Items can have up to three
+/// distinct marks by default, with each mark independently maintained. Users can use self-defined
+/// types as marks by implementing the `From<Type>` trait for [`XMark`]. Marking is also applicable
+/// to internal nodes, indicating marked descendant nodes, though such marking is not transparent
+/// to users.
+pub struct XArray<P, M = NoneMark>
+where
+    P: NonNullPtr + Send + Sync,
+{
+    head: RcuOption<XEntry<P>>,
```

Comment:
Possible to convert this to `RcuOption<XNode<P>>` since it seems the head entry is always a node?